test/unittests/Microsoft.ServiceFabric.Services.Wcf.Tests/WcfServiceRemotingClientTest.cs

```diff
@@ -0,0 +1,166 @@
+
+                Mock.Get(requestMessageMock)
+                    .Setup(m => m.GetHeader()) // We inject exception here for convenience (ideally, it should be in inner RequestResponseAsync call).
+                    .Throws(faultException);
+
+                // Act & Assert
+                // Assert that the exception is deserialized correctly and is of type NotImplementedException.
+                AggregateException exception = await Assert.ThrowsAsync<AggregateException>(() => sut.RequestResponseAsync(requestMessageMock));
+                Exception innerException = exception.Flatten().InnerException;
+                Assert.IsType<NotImplementedException>(innerException);
+                Assert.Equal(errorMessage, innerException.Message);
+            }
+        }
+
+        public sealed class UnsupportedException : WcfServiceRemotingClientTestBase
+        {
+            readonly FaultException<RemoteException2> faultException;
+            readonly string errorMessage;
+
+            public UnsupportedException()
+            {
+                // Create runtime exception convertors.
+                IEnumerable<V2.Runtime.IExceptionConvertor> runtimeExceptionConvertors = new List<V2.Runtime.IExceptionConvertor>
+                    {
+                        new CustomConvertorRuntime(),
+                    };
+
+                ExceptionSerializer exceptionSerializer = new ExceptionSerializer(
+                    runtimeExceptionConvertors,
+                    null
+                );
+
+                // Create RemoteException and FaultException
+                errorMessage = fuzzy.String();
+                RemoteException2 customRemoteException = exceptionSerializer.BuildRemoteException(new CustomException(errorMessage, "CustomField1", "CustomField2"));
+
+                faultException = new FaultException<RemoteException2>(customRemoteException);
+            }
+
+            [Fact]
+            public async Task RequestResponseAsync()
+            {
+                IServiceRemotingRequestMessage requestMessageMock = Mock.Of<IServiceRemotingRequestMessage>();
+
+                Mock.Get(requestMessageMock)
+                    .Setup(m => m.GetHeader()) // We inject exception here for convenience (ideally, it should be in inner RequestResponseAsync call).
+                    .Throws(faultException);
+
+                // Act & Assert
+                // Assert that the unknown exception is converted to ServiceException during deserialization.
+                AggregateException exception = await Assert.ThrowsAsync<AggregateException>(() => sut.RequestResponseAsync(requestMessageMock));
+                Exception innerException = exception.Flatten().InnerException;
+                Assert.IsType<ServiceException>(innerException);
+                Assert.Equal(errorMessage, innerException.Message);
+            }
+        }
+    }
+
+    internal class CustomConvertorRuntime : ExceptionConvertorBase
```

Comment:
Why are these top-level classes defined in this file? Looks like they aren't needed in other test classes, so it would be better to make them nested inside of the main test/base class. They'll be accessible by default and won't need to be exposed to the rest of the project.